propagation/propagation.go

```diff
@@ -23,97 +45,159 @@ import (
+// is assumed that supplier contains a valid honeycomb trace context header.
+// This is useful for testing.
+func (hc HoneycombContext) Unmarshal(ctx context.Context, supplier otelprop.HTTPSupplier) (*trace.SpanContext, error) {
+	ctx = hc.Extract(ctx, supplier)
+	sc := trace.GetRemoteSpanContextFromContext(ctx)
+	if sc == nil {
+		return nil, &PropagationError{"no remote span context found", nil}
 	}
-	return nil, &PropagationError{fmt.Sprintf("unrecognized version for trace header %s", getVer[0]), nil}
+	return sc, nil
 }
 
-// UnmarshalTraceContextV1 takes the trace header, stripped of the version
+// extractV1 takes the trace header, stripped of the version
 // string, and returns the component parts. Trace ID and Parent ID are both
 // required. If either is absent a nil trace header will be returned.
-func UnmarshalTraceContextV1(header string) (*Propagation, error) {
+func (HoneycombContext) extractV1(header string) (*trace.SpanContext, error) {
 	clauses := strings.Split(header, ",")
-	var prop = &Propagation{}
+	var sc = &trace.SpanContext{}
 	var tcB64 string
 	for _, clause := range clauses {
 		keyval := strings.SplitN(clause, "=", 2)
 		switch keyval[0] {
 		case "trace_id":
-			prop.TraceID = keyval[1]
+			sc.TraceID = keyval[1]
 		case "parent_id":
-			prop.ParentID = keyval[1]
+			sc.ParentID = keyval[1]
 		case "dataset":
-			prop.Dataset, _ = url.QueryUnescape(keyval[1])
+			sc.Dataset, _ = url.QueryUnescape(keyval[1])
 		case "context":
 			tcB64 = keyval[1]
 		}
 	}
-	if prop.TraceID == "" && prop.ParentID != "" {
+	if sc.TraceID == "" && sc.ParentID != "" {
 		return nil, &PropagationError{"parent_id without trace_id", nil}
 	}
 	if tcB64 != "" {
 		data, err := base64.StdEncoding.DecodeString(tcB64)
 		if err != nil {
 			return nil, &PropagationError{"unable to decode base64 trace context", err}
 		}
-		prop.TraceContext = make(map[string]interface{})
-		err = json.Unmarshal(data, &prop.TraceContext)
+		sc.TraceContext = make(map[string]interface{})
+		err = json.Unmarshal(data, &sc.TraceContext)
 		if err != nil {
 			return nil, &PropagationError{"unable to unmarshal trace context", err}
 		}
 	}
-	return prop, nil
+	return sc, nil
+
+}
+
+// Inject assembles the trace context header and sets it in the supplier.
+func (HoneycombContext) Inject(ctx context.Context, supplier otelprop.HTTPSupplier) {
+	sc := trace.GetRemoteSpanContextFromContext(ctx)
+	if sc == nil {
+		return
+	}
+	h := sc.SerializeHeader()
+	supplier.Set(HoneycombTracePropagationHTTPHeader, h)
+}
+
+// GetAllKeys returns the name of the header used for trace context propagation.
+func (HoneycombContext) GetAllKeys() []string {
+	return []string{HoneycombTracePropagationHTTPHeader}
+}
+
+// AmazonContext understands how to parse and generate Amazon trace propagation headers
+type AmazonContext struct{}
+
+// AmazonHTTPPropagator returns an HTTPPropagator that understands Amazon trace context
+// headers.
+func AmazonHTTPPropagator() trace.HTTPPropagator {
```

Comment:
Like `HoneycombContext` above, I recommend either returning type `AmazonContext` **or** unexporting the type.